cylc/flow/xtrigger_mgr.py

```diff
@@ -159,143 +159,70 @@ class TemplateVariables(Enum):
+        # Map xtrig label to function context.
         self.functx_map: 'Dict[str, SubFuncContext]' = {}
-        # When next to call a function, by signature.
-        self.t_next_call: dict = {}
-        # Satisfied triggers and their function results, by signature.
-        self.sat_xtrig: dict = {}
-        # Signatures of active functions (waiting on callback).
-        self.active: list = []
-
         # Clock labels, to avoid repeated string comparisons
         self.wall_clock_labels: Set[str] = set()
-
-        # Workflow wide default, used when not specified in xtrigger kwargs.
+        # Workflow-wide default, used when not specified in xtrigger kwargs.
         self.sequential_xtriggers_default = False
         # Labels whose xtriggers are sequentially checked.
         self.sequential_xtrigger_labels: Set[str] = set()
-        # Gather parentless tasks whose xtrigger(s) have been satisfied
-        # (these will be used to spawn the next occurrence).
-        self.sequential_spawn_next: Set[str] = set()
-        self.sequential_has_spawned_next: Set[str] = set()
 
-        self.workflow_run_dir = workflow_run_dir
+    def update(self, xtriggers: 'XtriggerCollator'):
+        self.functx_map.update(xtriggers.functx_map)
+        self.wall_clock_labels.update(xtriggers.wall_clock_labels)
+        self.sequential_xtrigger_labels.update(
+            xtriggers.sequential_xtrigger_labels)
 
-        # For function arg templating.
-        if not user:
-            user = get_user()
-        self.farg_templ: Dict[str, Any] = {
-            TemplateVariables.Workflow.value: workflow,
-            TemplateVariables.UserName.value: user,
-            TemplateVariables.RunDir.value: workflow_run_dir,
-            TemplateVariables.ShareDir.value: workflow_share_dir,
-            TemplateVariables.DebugMode.value: cylc.flow.flags.verbosity > 1,
-            # deprecated
-            TemplateVariables.WorkflowName.value: workflow,
-            TemplateVariables.SuiteName.value: workflow,
-            TemplateVariables.SuiteRunDir.value: workflow,
-            TemplateVariables.SuiteShareDir.value: workflow,
-        }
+    def add_trig(self, label: str, fctx: 'SubFuncContext', fdir: str) -> None:
+        """Add a new xtrigger function.
 
-        self.proc_pool = proc_pool
-        self.workflow_db_mgr = workflow_db_mgr
-        self.broadcast_mgr = broadcast_mgr
-        self.data_store_mgr = data_store_mgr
-        self.do_housekeeping = False
+        Args:
+            label: xtrigger label
+            fctx: function context
+            fdir: module directory
 
-    @staticmethod
-    def check_xtrigger(
+        """
+        if label in self.functx_map:
+            # we've already seen this one
+            return
+
+        if (
+            not label.startswith('_cylc_retry_') and not
+            label.startswith('_cylc_submit_retry_')
+        ):
+            # (the "_wall_clock" function fails "wall_clock" validation)
```

Comment:
?

Reply:
Have a look at `xtriggers/wall_clock.py` - It's wierd.

Reply:
I know, but I don't see what that has to do with:
- validation, as `_wall_clock` does not go through any validation process.
- the `if` condition above.

Reply:
The "weirdness" allows us to avoid recomputing the offset from cycle point every time a clock trigger is called. We do it once, and then call `_wall_clock` instead with an absolute time.

> the if condition above.

retries timers are implemented via wall_clock xtriggers

> validation

(This was needed, but I need to remind myself of why...)